torch/testing/_internal/common_methods_invocations.py

```diff
@@ -2089,6 +2112,28 @@ def _make_tensor_helper(shape, low=None, high=None):
                         # see discussion https://github.com/pytorch/pytorch/pull/47761#issuecomment-747316775
                         SkipInfo('TestGradients', 'test_fn_gradgrad', device_type='cuda'),)
                     ),
+    OpInfo('linalg.det',
+           op=torch.linalg.det,
+           dtypes=floating_and_complex_types(),
+           aten_name='linalg_det',
+           sample_inputs_func=sample_inputs_linalg_det,
+           decorators=[skipCUDAIfNoMagma, skipCPUIfNoLapack],
+           test_complex_grad=False,
+           test_inplace_grad=False,
+           skips=(
+               # https://github.com/pytorch/pytorch/issues/53361
```

Comment:
Let's file a new issue specifically for torch.linalg.det's autograd failure

Reply:
I wrote in the issue that the analysis that the autograd test for det was never run on cuda is not correct https://github.com/pytorch/pytorch/issues/53361#issuecomment-791588410. The reason for this fail https://app.circleci.com/pipelines/github/pytorch/pytorch/281078/workflows/bd828f67-1892-45ba-81cf-164b352ead16/jobs/11315029 most probably is not the error in the implementation. My guess that the problem is not in backward implementation is supported by the fact that only gradgradcheck fails and gradcheck passes.
Looking at the Jacobian values my best guess that the problem is with the numerical error of finite differences to estimate the "numerical Jacobian".

We shouldn't skip it, the test was working before. Can you tell for which SampleInput specifically the test fails?

Reply:
Thanks for the analysis @IvanYashchuk, that's really interesting.

Reply:
I disagree with the analysis (or better, I agree with it but it doesn't apply in this specific case).
I wrote up a detailed explanation here: https://github.com/pytorch/pytorch/issues/53361#issuecomment-809493569

TL;DR: `test_gradgrad` does NOT convert inputs to the correct device/dtype. The error that we see here is present also on master, if you put the correct device in the sample inputs.

That said, I also think that the implementation is correct and that the error that we see here is just an artifact produced by gradcheck. I admit I have no idea how to solve though, apart from skipping it.

Also, I don't think that this should block the merge of the PR since it's not a regression. On the contrary, it is an improvement over master, were `torch.det` autograd on cuda is completely untested.

Reply:
The real problem is that for inputs with a rank less than `n-1` the backward is unstable.

<details>
  <summary>Here is the code that demonstrates the failure on CPU.  Setting seed to 4 the tests fail, setting seed to 5 the tests pass.</summary>


```python
import torch
import numpy as np
from torch.testing._internal.common_utils import random_square_matrix_of_rank
from torch.autograd.gradcheck import gradcheck, gradgradcheck

torch.manual_seed(4) # with seed 5 gradgradcheck passes

S = 3
dtype = torch.float64
device = 'cpu'

kw = dict(device=device, dtype=dtype)
inputs = [
    random_square_matrix_of_rank(S, S - 2, **kw),
]

print(inputs)

A = inputs[0].clone()
A.requires_grad_()

def svd_based_det(x):
  _, s, _ = torch.svd(x)
  return torch.prod(s)


gradcheck(torch.linalg.det, A) # this passes
gradgradcheck(torch.linalg.det, A) # this fails

gradcheck(svd_based_det, A) # this passes
gradgradcheck(svd_based_det, A) # this fails
```

</details>

Please comment out dim2_null, rank1, rank2 sample_inputs with a comment pointing to the issue about backward stability of det https://github.com/pytorch/pytorch/issues/53364.
Then this skip of `test_fn_gradgrad` can be removed.

Reply:
wow, thank you very much for the analysis. I implemented your suggestions in b71452856f